extensions/amp-autocomplete/0.1/amp-autocomplete.js

```diff
@@ -14,43 +14,415 @@
+    element.setAttribute('role', 'listitem');
+    element.textContent = item;
+    return element;
+  }
+
+  /**
+  * Handle rendering results on user input.
+  * @param {Event} event
+  * @return {!Promise}
+  * @private
+  */
+  inputHandler_(event) {
+    // Consider caching results to avoid repetitive queries on delete.
+    if (!event.inputType) {
+      return Promise.resolve();
+    }
+    return this.mutateElement(() => {
+      this.renderResults_();
+    }).then(() => {
+      this.showResults();
+    });
+  }
+
+  /**
+   * Handle selecting items on user mousedown.
+   * @param {Event} event
+   * @return {!Promise}
+   * @private
+   */
+  selectHandler_(event) {
+    if (!this.isItemElement(event.target)) {
+      return Promise.resolve();
+    }
+    return this.mutateElement(() => {
+      this.selectItem(event.target);
+    });
+  }
+
+  /**
+   * Render filtered results on the current input and update the container_.
+   * @private
+   */
+  renderResults_() {
+    const userInput = this.inputElement_.value;
+    this.clearAllItems();
+    if (userInput.length < this.minChars_ || !this.inlineData_) {
+      return;
+    }
+    const filteredData = this.filterData_(this.inlineData_, userInput);
+    filteredData.forEach(item => {
+      this.container_.appendChild(this.createElementFromItem_(item));
+    });
+  }
+
+  /**
+   * Apply the filter to the given data based on the given input.
+   * @param {!Array<string>} data
+   * @param {string} input
+   * @return {!Array<string>}
+   * @private
+   */
+  filterData_(data, input) {
+    let filteredData = data.filter(item => {
+      switch (this.filter_) {
+        case FilterType.SUBSTRING:
+          return includes(item, input);
+        case FilterType.PREFIX:
+          return startsWith(item, input);
+        case FilterType.TOKEN_PREFIX:
+          return item.split(' ').some(token => {
+            return startsWith(token, input);
+          });
+        case FilterType.FUZZY:
+          throw new Error(`Filter not yet supported: ${this.filter_}`);
+        case FilterType.CUSTOM:
+          throw new Error(`Filter not yet supported: ${this.filter_}`);
+        case FilterType.NONE:
+          // Query server endpoint.
+          throw new Error(`Filter not yet supported: ${this.filter_}`);
+        default:
+          throw new Error(`Unexpected filter: ${this.filter_}`);
+      }
+    });
+
+    // Truncate to max-entries.
+    if (this.maxEntries_ && this.maxEntries_ < filteredData.length) {
+      filteredData = filteredData.slice(0, this.maxEntries_);
+    }
+
+    return filteredData;
+  }
+
+  /** Set container_ visibility to visible. */
+  showResults() {
+    if (!this.container_) {
+      return;
+    }
+    toggle(this.container_, true);
+  }
+
+  /** Set container_ visibility to hidden.
+   * @return {!Promise}
+   */
+  hideResults() {
+    if (!this.container_) {
+      return Promise.resolve();
+    }
+    toggle(this.container_, false);
```

Comment:
`toggle` calls need to be inside a `mutateElement` block, since `toggle` applies the `hidden` attribute to the given element, and the `hidden` attribute changes an element's CSS `display` property, which causes reflow https://csstriggers.com/display.